From: Federico Pellegrin <fede@evolware.org>
Date: Sat, 29 Jan 2022 10:36:36 +0100
Subject: [PATCH] gpmi-nand fix MX6 clock bug from upstream

Apply patches to fix timing problems, seen with some Micron chips, see:

https://git.phytec.de/linux-mainline/commit/?h=v5.10.48-phy&id=866939ea8110764d9c12af960d746e2f7f5debe3

and:

https://github.com/torvalds/linux/commit/aa1baa0e6c1aa4872e481dce4fc7fd6f3dd8496b#diff-0dec2fa8640ea2067789c406ab1e42c9805d0d0fc9f70a3a29d17f9311e23ca2

and complete discussions at:

https://lore.kernel.org/all/4352240.LvFx2qVVIh@n95hx1g2/

diff -ruN a/drivers/mtd/nand/raw/gpmi-nand/gpmi-nand.c b/drivers/mtd/nand/raw/gpmi-nand/gpmi-nand.c
--- a/drivers/mtd/nand/raw/gpmi-nand/gpmi-nand.c	2021-10-30 11:27:44.872933754 +0200
+++ b/drivers/mtd/nand/raw/gpmi-nand/gpmi-nand.c	2022-01-29 10:09:34.176511613 +0100
@@ -710,15 +710,25 @@
 			      (use_half_period ? BM_GPMI_CTRL1_HALF_PERIOD : 0);
 }
 
-static void gpmi_nfc_apply_timings(struct gpmi_nand_data *this)
+static int gpmi_nfc_apply_timings(struct gpmi_nand_data *this)
 {
 	struct gpmi_nfc_hardware_timing *hw = &this->hw;
 	struct resources *r = &this->resources;
 	void __iomem *gpmi_regs = r->gpmi_regs;
 	unsigned int dll_wait_time_us;
+	int ret;
+
+	ret = __gpmi_enable_clk(this, false);
+	if (ret)
+		return ret;
 
 	clk_set_rate(r->clock[0], hw->clk_rate);
 
+	ret = __gpmi_enable_clk(this, true);
+	if (ret)
+		return ret;
+
+
 	writel(hw->timing0, gpmi_regs + HW_GPMI_TIMING0);
 	writel(hw->timing1, gpmi_regs + HW_GPMI_TIMING1);
 
@@ -736,6 +746,8 @@
 
 	/* Wait for the DLL to settle. */
 	udelay(dll_wait_time_us);
+
+	return 0;
 }
 
 static int gpmi_setup_data_interface(struct nand_chip *chip, int chipnr,
@@ -1184,15 +1196,6 @@
 		r->clock[i] = clk;
 	}
 
-	if (GPMI_IS_MX6(this))
-		/*
-		 * Set the default value for the gpmi clock.
-		 *
-		 * If you want to use the ONFI nand which is in the
-		 * Synchronous Mode, you should change the clock as you need.
-		 */
-		clk_set_rate(r->clock[0], 22000000);
-
 	return 0;
 
 err_clock:
@@ -2429,7 +2432,9 @@
 	 */
 	if (this->hw.must_apply_timings) {
 		this->hw.must_apply_timings = false;
-		gpmi_nfc_apply_timings(this);
+		ret = gpmi_nfc_apply_timings(this);
+		if (ret)
+			goto unmap;
 	}
 
 	dev_dbg(this->dev, "%s: %d instructions\n", __func__, op->ninstrs);
